File: backend-v2/src/zhishi/server/routes/focus.py
# src/zhishi/server/routes/focus.py
from datetime import datetime

from fastapi import APIRouter, Depends, HTTPException, status
from pydantic import BaseModel, ConfigDict
from sqlalchemy.orm import Session
from zhishi.domain.focus import service
from zhishi.domain.focus.schemas import TimerStart
from zhishi.domain.models import TimeLog
from zhishi.server.deps import get_db

router = APIRouter(prefix="/api/focus", tags=["focus"])


# ---- typed 响应（re #047：前端 FocusLog/FocusStats 手写收敛依据；
# started_at/ended_at 序列化为本地 naive ISO 串，与既有回包一致） ----

class TimeLogOut(BaseModel):
    id: int
    task_id: int | None = None
    task_title: str
    kind: str                    # focus/break
    started_at: datetime
    ended_at: datetime | None = None
    minutes: int


class FocusStopMissOut(BaseModel):
    """stop 落空（无进行中计时或目标已结账）的回包；extra=forbid 保证
    stop 端点 TimeLogOut | FocusStopMissOut 二选一互不误配。"""
    model_config = ConfigDict(extra="forbid")
    ok: bool = False
    stopped: None = None


class ByDayItem(BaseModel):
    date: str                    # YYYY-MM-DD
    minutes: int


class ByTaskItem(BaseModel):
    task_title: str
    minutes: int


class FocusStatsOut(BaseModel):
    by_day: list[ByDayItem]
    by_task: list[ByTaskItem]
    total_minutes: int


@router.post("/start", status_code=201, response_model=TimeLogOut)
def start(payload: TimerStart, db: Session = Depends(get_db)):
    return _log_dict(service.start_timer(db, payload))


@router.post("/stop", response_model=TimeLogOut | FocusStopMissOut)
def stop(body: dict | None = None, db: Session = Depends(get_db)):
    log_id = body.get("log_id") if body else None
    log = service.stop_timer(db, log_id)
    return _log_dict(log) if log is not None else {"ok": False, "stopped": None}


@router.get("/current", response_model=TimeLogOut | None)
def current(db: Session = Depends(get_db)):
    log = service.current_log(db)
    return _log_dict(log) if log is not None else None


@router.get("/logs", response_model=list[TimeLogOut])
def logs(days: int = 7, task_id: int | None = None, db: Session = Depends(get_db)):
    return [_log_dict(l) for l in service.list_logs(db, days=days, task_id=task_id)]


@router.get("/stats", response_model=FocusStatsOut)
def stats(days: int = 7, db: Session = Depends(get_db)):
    return service.time_stats(db, days=days)


@router.delete("/logs/{log_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_log(log_id: int, db: Session = Depends(get_db)) -> None:
    """删除一条已结束的计时记录（re k3#048 残留清理诉求）。
    运行中的计时不可直接删（先停后删），409 防误删破坏 current 指针语义。"""
    log = db.get(TimeLog, log_id)
    if log is None:
        raise HTTPException(404, "计时记录不存在")
    if log.ended_at is None:
        raise HTTPException(409, "运行中的计时不能删除，请先停止")
    db.delete(log)
    db.commit()


def _log_dict(l):
    return {"id": l.id, "task_id": l.task_id, "task_title": l.task_title, "kind": l.kind,
            "started_at": l.started_at, "ended_at": l.ended_at, "minutes": l.minutes}

File: frontend-v2/app/src/api/focus.ts
/**
 * 番茄钟（专注计时）客户端。
 * 契约：/api/focus 系列响应已 typed（TimeLogOut/FocusStatsOut/FocusStopMissOut），
 * 日志/统计/停钟未命中均派生自 contracts 生成类型，不再手写字段。已核实行为：
 * - 开始：POST /api/focus/start（201），body TimerStart{task_id?, task_title='', kind='focus'|'break'} → log
 * - 结束：POST /api/focus/stop，body {log_id?}（缺省停当前）→ log；无进行中 → {"ok": false, "stopped": null}
 * - 当前：GET /api/focus/current → log | null（再 start 会顶替 current 指向最新一条）
 * - 历史：GET /api/focus/logs?days=7&task_id= → log[]
 * - 删除：DELETE /api/focus/logs/{log_id} → 204；409 = 该条仍在进行中不可删
 * - 统计：GET /api/focus/stats?days=7 → {by_day:[{date,minutes}], by_task:[{task_title,minutes}], total_minutes}
 * started_at/ended_at 为本地 naive ISO 字符串（无时区后缀），new Date 按本地时区解析即与后端一致。
 */
import type { components } from './contracts/rest'
import { http } from './http'

export type FocusKind = 'focus' | 'break'

/** 计时记录 = 生成 TimeLogOut。kind 后端为自由串，实际取值 focus|break。 */
export type FocusLog = components['schemas']['TimeLogOut']

/** 统计 = 生成 FocusStatsOut（by_day/by_task 项见 ByDayItem/ByTaskItem）。 */
export type FocusStats = components['schemas']['FocusStatsOut']

/** POST /stop 时无进行中计时的回包 = 生成 FocusStopMissOut（有则直接返回结账后的 FocusLog）。 */
export type FocusStopMiss = components['schemas']['FocusStopMissOut']

export interface FocusStartBody {
  task_id?: number | null
  task_title?: string
  kind?: FocusKind
}

export function startFocus(body: FocusStartBody): Promise<FocusLog> {
  return http.post('/api/focus/start', body)
}

export function stopFocus(logId?: number): Promise<FocusLog | FocusStopMiss> {
  return http.post('/api/focus/stop', logId === undefined ? {} : { log_id: logId })
}

export function getCurrentFocus(): Promise<FocusLog | null> {
  return http.get('/api/focus/current')
}

export function listFocusLogs(days = 7, taskId?: number): Promise<FocusLog[]> {
  return http.get('/api/focus/logs', { days, task_id: taskId })
}

/** 删除一条计时记录（204）；该条仍在进行中时后端 409（按 http 层惯例抛 HttpError，交给 store 捕获）。 */
export function deleteFocusLog(logId: number): Promise<void> {
  return http.del(`/api/focus/logs/${logId}`)
}

export function getFocusStats(days = 7): Promise<FocusStats> {
  return http.get('/api/focus/stats', { days })
}
